src/dialog/llm/lcel_default.py

```diff
@@ -0,0 +1,113 @@
+import asyncio
+import logging
+
+from langchain.chains.llm import LLMChain
+from langchain.memory import ConversationBufferWindowMemory
+from langchain.memory.chat_memory import BaseChatMemory
+from langchain.prompts import (
+    ChatPromptTemplate,
+    HumanMessagePromptTemplate,
+    SystemMessagePromptTemplate
+)
+from langchain_openai.chat_models import ChatOpenAI
+
+from dialog.learn.idf import categorize_conversation_history
+from dialog.llm.abstract_llm import AbstractLLM
+from dialog.llm.embeddings import get_most_relevant_contents_from_message
+from dialog.llm.memory import generate_memory_instance
+from dialog.settings import (
+    LLM_MEMORY_SIZE,
+    LLM_RELEVANT_CONTENTS,
+    OPENAI_API_KEY,
+    VERBOSE_LLM
+)
+
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.prompts import ChatPromptTemplate
+
+
+class DialogLcelLLM(AbstractLLM):
+    parser = StrOutputParser()
+
+    @property
+    def memory(self) -> BaseChatMemory:
+        if self.session_id:
+            return generate_memory_instance(
+                session_id=self.session_id,
+                parent_session_id=self.parent_session_id
+            )
+        return None
+
+    def generate_prompt(self, text):
+        self.relevant_contents = get_most_relevant_contents_from_message(
+            text, top=LLM_RELEVANT_CONTENTS, dataset=self.dataset
+        )
+        prompt_config = self.config.get("prompt")
+        fallback = prompt_config.get("fallback") or self.config.get("fallback").get("prompt") # maintaining compatibility with the previous configuration
+        header = prompt_config.get("header")
+        suggested = prompt_config.get("suggested")
+        messages = []
+        if len(self.relevant_contents) > 0:
+            context = "Context: \n".join(
+                [f"{c.question}\n{c.content}\n" for c in self.relevant_contents]
+            )
+            self.prompt = ChatPromptTemplate.from_template(
+                "\n".join([
+                    SystemMessagePromptTemplate.from_template(header),
```

Comment:
this is breaking because the `join` over the list expects strings, not the langchain template object:
```
dialog-1  | INFO:     192.168.65.1:47093 - "POST /ask HTTP/1.1" 500 Internal Server Error
dialog-1  | ERROR:    Exception in ASGI application
dialog-1  | Traceback (most recent call last):
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
dialog-1  |     result = await app(  # type: ignore[func-returns-value]
dialog-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
dialog-1  |     return await self.app(scope, receive, send)
dialog-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
dialog-1  |     await super().__call__(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
dialog-1  |     await self.middleware_stack(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
dialog-1  |     raise exc
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
dialog-1  |     await self.app(scope, receive, _send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/cors.py", line 83, in __call__
dialog-1  |     await self.app(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
dialog-1  |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
dialog-1  |     raise exc
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
dialog-1  |     await app(scope, receive, sender)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 758, in __call__
dialog-1  |     await self.middleware_stack(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 778, in app
dialog-1  |     await route.handle(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 299, in handle
dialog-1  |     await self.app(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 79, in app
dialog-1  |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
dialog-1  |     raise exc
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
dialog-1  |     await app(scope, receive, sender)
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 74, in app
dialog-1  |     response = await func(request)
dialog-1  |                ^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 299, in app
dialog-1  |     raise e
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 294, in app
dialog-1  |     raw_response = await run_endpoint_function(
dialog-1  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
dialog-1  |     return await dependant.call(**values)
dialog-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/app/src/main.py", line 98, in ask_question_to_llm
dialog-1  |     ai_message = llm_instance.process(message.message)
dialog-1  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dialog-1  |   File "/app/src/dialog/llm/lcel_default.py", line 108, in process
dialog-1  |     self.generate_prompt(processed_input)
dialog-1  |   File "/app/src/dialog/llm/lcel_default.py", line 55, in generate_prompt
dialog-1  |     "\n".join([
dialog-1  | TypeError: sequence item 0: expected str instance, SystemMessagePromptTemplate found
```
it would be better to continue to use langchain system and human templates like we already do in the [default llm](https://github.com/talkdai/dialog/blob/feature/lcel-support/src/dialog/llm/default.py#L30) or get some ideas from [langchain's Q&A documentation](https://python.langchain.com/docs/use_cases/question_answering/chat_history)

Reply:
Fixing this.